Show explanation if no files are found

esanzgar opened the pull request:
Resolves #765

![image](https://user-images.githubusercontent.com/8555781/94166988-4681a700-fe8c-11ea-84b7-518586781c30.png)

esanzgar commented:
@klemay when you know how the message should look like let me know.

klemay commented:
@esanzgar suggested copy: 

There are no supported files in the current Canvas course. Please see this Canvas help article for [uploading files to a course](https://community.canvaslms.com/t5/Instructor-Guide/How-do-I-upload-a-file-to-a-course/ta-p/618) and try again.

esanzgar commented:
@klemay 

Is this acceptable?

![image](https://user-images.githubusercontent.com/8555781/94249807-b3468100-ff20-11ea-9d1c-134742cd3095.png)

esanzgar commented:
Latest look and feel:
![image](https://user-images.githubusercontent.com/8555781/94413735-c1dca480-017b-11eb-853d-f689aa6d5d55.png)

robertknight commented on lms/static/styles/components/_FileList.scss:
```diff
@@ -58,3 +60,11 @@ $file-list-icon-size: 24px;
+  height: 40px;
+  vertical-align: middle;
+  padding: 12px;
+  background-color: #eee;
```
Perhaps use `var.$grey-2` here (#ececec) for consistency with the disabled buttons in the form. Using color variables for foreground and background colors also makes it easier to check that contrast is high enough everywhere.

esanzgar replied:
I was making it consistent with the look and feel of the zebra-stripe rows in the normal table (https://github.com/hypothesis/lms/blob/c0fa85c9a79186b3efda91e82a15ce5fb763350b/lms/static/styles/lms.css#L193).

Sure, I will make it `var.$grey-2`

robertknight commented on lms/static/scripts/frontend_apps/components/FileList.js:
```diff
@@ -71,6 +71,24 @@ export default function FileList({
+              uploading files to a course
+            </a>{' '}
+            and try again.
+          </p>
```
The text looks good. This text is specific to a particular LMS whereas the `FileList` component itself is currently not. One change that we plan to make soon is to re-use this component with other LMSes such as Blackboard.

What I would suggest is that we keep the basic structure here, but pass in the content of the `<div>` and the `<p>` as props from the parent `LMSFilePicker` component. The parent can then pass in the appropriate message text depending on the LMS.

esanzgar replied:
I implemented your suggestion.

Your approach is different from [Sean's proposal](https://github.com/hypothesis/lms/pull/2100) which make the FileList less agnostic.

robertknight reviewed (changes requested):
Hi Eduardo,

Visually it looks good. One change I would like to make to prepare for a feature that we plan to work on soon is to keep the `FileList` component agnostic of the current LMS and pass in the LMS-specific messages as props from the parent `LMSFilePicker` component.

esanzgar commented on lms/static/scripts/frontend_apps/components/test/FileList-test.js:
```diff
@@ -60,6 +63,11 @@ describe('FileList', () => {
     assert.isFalse(wrapper.exists('.FileList__spinner'));
   });
 
+  it('renders a explanatory message when ', () => {
```
Done

esanzgar commented on lms/static/styles/components/_FileList.scss:
```diff
@@ -38,7 +40,7 @@ $file-list-icon-size: 24px;
     // nb. Argument must be passed to `invert` even though "1" is the default
     // to work around https://bugs.chromium.org/p/chromium/issues/detail?id=964696.
-    filter: #{"invert(1)"}; // Quoted to avoid conflict with SASS mixin.
+    filter: #{'invert(1)'}; // Quoted to avoid conflict with SASS mixin.
```
Unintentional.

Because .prettierrc has `"singleQuote": true` it changed the quote style when saving the file.

Is this something I should avoid in the future?

robertknight replied:
> Because .prettierrc has "singleQuote": true it changed the quote style when saving the file.

Ah. In the lms repo prettier currently only applies to JavaScript. See the `format` command in `package.json`. We already use Prettier to format .scss files in the client repo so it would make sense to use it here as well. See `package.json` in the client repo. If we're going to make that change it should be in a new PR which modifies the "format" and "checkformatting" scripts and also reformats the existing `.scss` files.

esanzgar commented on lms/static/scripts/frontend_apps/components/FileList.js:
```diff
@@ -6,6 +6,7 @@ import Table from './Table';
 
 /**
  * @typedef {import('../api-types').File} File
+ * @typedef {import("preact").VNode} VNode
```
I use VNode because it allows string, JSX components, and arrays of components.

esanzgar commented:
In the latest guidelines from @klemay, she requested to remove the 'Select' button when no files were available.

 Like this:
![image](https://user-images.githubusercontent.com/8555781/95453304-e0585200-096a-11eb-8d44-54209e41aad8.png)

In order to remove the buttons in a more cleaner way, I introduced a second commit  (from [this draft PR](https://github.com/hypothesis/lms/pull/2147)) which added more flexibility to the `LMSFilePicker` component (to add or remove titles and buttons in a simpler way). 

A regression comment about the second commit. There is a small deviation from the previous behaviour: when the fetching of files fails generically (not because of authorization issues), it retries the fetching without bugging the user to re-authorise the access to files.

robertknight reviewed (commented):
Hi Eduardo,

Everything worked as expected in my manual testing. The code looked good. I requested some minor changes to a few details.

The main query I have is about the change to the behaviour when an error occurs with an associated error message. Can you remind me what prompted that change, as it isn't mentioned in the PR description?

esanzgar commented on lms/static/scripts/frontend_apps/components/LMSFilePicker.js:
```diff
@@ -136,62 +175,82 @@ export default function LMSFilePicker({
+        />,
+      ];
+      warningError = (
+        <p label="authorization warning">
```
You are correct. I will change them.

esanzgar commented on lms/static/scripts/frontend_apps/components/LMSFilePicker.js:
```diff
@@ -136,62 +175,82 @@ export default function LMSFilePicker({
-  const title =
-    dialogState.state === 'authorizing' ? 'Allow file access' : 'Select a file';
+  let buttons;
+  let warningError;
```
I renamed to `warningOrError` to make it more generic, in case it needed to be extended to other type of warnings or errors. I hope it is OK.

esanzgar commented:
In response to @robertknight:

In order to remove the ‘Select’ file button I refactored the code to make it more flexible. During this refactor I realised that if the API returned an error not related to authorisation, the dialog warning and the action was authorisation related. That did not seem right to me. For example, in case of a network error during the API request, I think it is not suitable to display a warning about authorisation. Instead, I changed to display a generic warning and to retry the fetching of files. It only prompted the authorisation dialogue when the error is specific about authorisation.

However, as @robertknight mentioned, re-authorizing _maybe_ the right thing to do, therefore, I changed a line in the code so the behaviour is equivalent to the previous version. In the test there is a hint that may indicate the re-authorizing could be the right behaviour.

robertknight commented on lms/static/scripts/frontend_apps/components/test/LMSFilePicker-test.js:
```diff
@@ -114,8 +114,12 @@ describe('LMSFilePicker', () => {
+  it('shows the "Authorize" and "Authorize again" buttons after 2 failed authorization requests', async () => {
+    fakeApiCall.rejects(
+      new ApiError('Not authorized', {
+        /** withouth message */
```
```suggestion
        /** withouth message */
```

Should this comment say something like "without `errorMessage` property"?

esanzgar replied:
Yes, it should be.

robertknight commented on lms/static/scripts/frontend_apps/components/test/LMSFilePicker-test.js:
```diff
@@ -201,6 +236,38 @@ describe('LMSFilePicker', () => {
     });
   });
 
+  it('shows no buttons when the request returns an empty list of files', async () => {
```
The dialog's built-in "Cancel" button will still be shown in this case. Perhaps "shows no continue button ..." or "shows no select button ...".

robertknight commented on lms/static/scripts/frontend_apps/components/LMSFilePicker.js:
```diff
@@ -74,24 +97,45 @@ export default function LMSFilePicker({
   // Execute the authorization flow in a popup window and then attempt to
   // fetch files.
   const authorizeAndFetchFiles = useCallback(async () => {
-    setDialogState({ ...INITIAL_DIALOG_STATE, state: 'authorizing' });
```
There are still changes to the `authorizeAndFetchFiles` logic here. Are these still applicable or do they need to be reverted?

esanzgar replied:
No, they don't need to be reverted.

The 'authorizing' state is set within the `fetchFiles` function, when an error is raised from the API.

robertknight commented on lms/static/scripts/frontend_apps/components/LMSFilePicker.js:
```diff
@@ -118,11 +161,7 @@ export default function LMSFilePicker({
-    } else {
-      fetchFiles();
-    }
+    fetchFiles();
```
Is this change still applicable given that we're reverting the authorization error-handling changes in this PR?

esanzgar replied:
This useEffect hook is only run when the component is mounted. At that time, the `dialogState.state` is never 'authorizing'.

This is my understanding on these lines. That's why I remove the lines from 121 to 125.

robertknight replied:
This makes sense. Yesterday I missed the fact that `dialogState.state` is always `fetching` here.

robertknight reviewed (commented):
I had a couple of queries about the changes to `authorizeAndFetchFiles`.

robertknight reviewed (approved):
LGTM 👍
